File: src/main/java/info/opensigma/ui/clickgui/frame/Frame.java
package info.opensigma.ui.clickgui.frame;

import info.opensigma.system.IMinecraft;
import info.opensigma.util.math.GeometryUtils;
import net.minecraft.client.gui.DrawableHelper;
import net.minecraft.client.util.math.MatrixStack;

public abstract class Frame extends DrawableHelper implements IMinecraft {

    protected double posX, posY;

    private boolean dragging;
    private double dragX, dragY;

    public final void draw(final MatrixStack matrices, final int mouseX, final int mouseY) {
        if (dragging) {
            posX = mouseX - dragX;
            posY = mouseY - dragY;
        }

        drawFrame(matrices, mouseX, mouseY);
    }

    protected abstract void drawFrame(final MatrixStack matrices, final int mouseX, final int mouseY);

    public final boolean mouseClick(final double mouseX, final double mouseY, int button) {
        if (GeometryUtils.isInBounds(mouseX, mouseY, posX, posY, getInteractionBounds()[0], getInteractionBounds()[1], true)) {
            dragX = mouseX - posX;
            dragY = mouseY - posY;
            dragging = true;

            return true;
        }

        return mouseClickFrame(mouseX, mouseY, button);
    }

    protected abstract boolean mouseClickFrame(final double mouseX, final double mouseY, final int button);

    public final boolean mouseRelease(final double mouseX, final double mouseY, final int button) {
        if (dragging) {
            dragging = false;
            dragX = 0;
            dragY = 0;

            return true;
        }

        return mouseReleaseFrame(mouseX, mouseY, button);
    }

    protected abstract boolean mouseReleaseFrame(final double mouseX, final double mouseY, final int button);

    protected abstract double[] getInteractionBounds();

}

File: src/main/kotlin/info/opensigma/ui/clickgui/frame/Frame.kt
package info.opensigma.ui.clickgui.frame

import info.opensigma.system.IMinecraft
import info.opensigma.util.math.GeometryUtils
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.components.Renderable

abstract class Frame : Renderable, IMinecraft {

    open var posX = 0.0
    open var posY = 0.0

    private var dragging = false
    private var dragX = 0.0
    private var dragY = 0.0

    override fun render(context: GuiGraphics, mouseX: Int, mouseY: Int, delta: Float) {
        if (dragging) {
            posX = mouseX - dragX
            posY = mouseY - dragY
        }

        drawFrame(context, mouseX, mouseY, delta)
    }

    protected abstract fun drawFrame(context: GuiGraphics, mouseX: Int, mouseY: Int, delta: Float)

    fun mouseClick(mouseX: Double, mouseY: Double, button: Int): Boolean {
        if (GeometryUtils.isInBounds(mouseX, mouseY, posX, posY, getInteractionBounds()[0], getInteractionBounds()[1], true)) {
            dragX = mouseX - posX
            dragY = mouseY - posY
            dragging = true

            return true
        }

        return mouseClickFrame(mouseX, mouseY, button)
    }

    @Suppress("SameReturnValue")
    protected abstract fun mouseClickFrame(mouseX: Double, mouseY: Double, button: Int): Boolean

    fun mouseRelease(mouseX: Double, mouseY: Double, button: Int): Boolean {
        if (dragging) {
            dragging = false
            dragX = 0.0
            dragY = 0.0

            return true
        }

        return mouseReleaseFrame(mouseX, mouseY, button)
    }

    @Suppress("SameReturnValue")
    protected abstract fun mouseReleaseFrame(mouseX: Double, mouseY: Double, button: Int): Boolean

    protected abstract fun getInteractionBounds(): DoubleArray

}